lib/charms/rolling_ops/v0/rollingops.py

```diff
@@ -0,0 +1,353 @@
+        app_status = self.relation.data[self.app].get(self.unit, self.IDLE)
+
+        if app_status == self.GRANTED and unit_status == self.RELEASE:
+            # Active release request.
+            return self.RELEASE
+
+        if app_status == self.IDLE and unit_status == self.ACQUIRE:
+            # Active acquire request.
+            return self.ACQUIRE
+
+        return app_status  # Granted or unset/released
+
+    @status.setter
+    def status(self, status):
+        """Set the given status.
+
+        Since we update the relation data, this may fire off a RelationChanged event.
+        """
+        if status == self.ACQUIRE:
+            self.relation.data[self.unit].update({"status": status})
+
+        if status == self.RELEASE:
+            self.relation.data[self.unit].update({"status": status})
+
+        if status == self.GRANTED:
+            self.relation.data[self.app].update({self.unit: status})
+
+        if status is self.IDLE:
+            self.relation.data[self.app].update({self.unit: status})
+
+    def acquire(self):
+        self.status = self.ACQUIRE
+
+    def release(self):
+        self.status = self.RELEASE
+
+    def clear(self):
+        self.status = self.IDLE
+
+    def grant(self):
+        self.status = self.GRANTED
+
+    def is_held(self):
+        return self.status == self.GRANTED
+
+    def release_requested(self):
+        return self.status == self.RELEASE
+
+    def is_pending(self):
+        return self.status == self.ACQUIRE
+
+
+class Locks:
+    """Generator that returns a list of locks."""
+    def __init__(self, rel_name, charm):
+        self.rel_name = rel_name
+        self.charm = charm
+
+        # Gather all the units.
+        relation = charm.model.relations[rel_name][0]
+        units = [unit for unit in relation.units]
+
+        # Plus our unit ...
+        units.append(charm.framework.model.unit)
+
+        self.units = units
+
+    def __iter__(self):
+
+        for unit in self.units:
+            yield Lock(self.rel_name, self.charm, unit=unit)
+
+
+class RollingEvent(EventBase):
+    def __init__(self, handle, name):
+        super().__init__(handle)
+        self._name = name
+
+    def snapshot(self):
+        return {"name": self._name}
+
+    def restore(self, snapshot):
+        self._name = snapshot["name"]
+
+    @property
+    def name(self):
+        """The `name` property allows us to namespace Rolling Operations."""
+        return self._name
+
+
+class RunWithLock(RollingEvent):
+    pass
+
+
+class AcquireLock(RollingEvent):
+    pass
+
+
+class ProcessLocks(RollingEvent):
+    pass
+
+
+class RollingEvents(CharmEvents):
+    run_with_lock = EventSource(RunWithLock)
+    acquire_lock = EventSource(AcquireLock)
+    process_locks = EventSource(ProcessLocks)
+
+
+class RollingOpsManager(Object):
+    """Emitters and handlers for rolling ops."""
+
+    def __init__(self, charm, rel_name, func_run, event_acquire):
```

Comment:
So I dropped a comment up top about this, but I think we should consider the following signature here:

```python
def __init__(self, charm, *, relation, trigger, callback):
```

To me at least, the language `trigger` and `callback` more obviously conveys the intent here? Open to discussion, just my two cents! :)

Reply:
That makes sense. I think that I was trying to avoid "callback" because this is almost, but not quite a callback. But it reads a lot cleaner, that way.